parquet/src/main/java/org/apache/iceberg/parquet/ParquetTypeWithPartnerVisitor.java

```diff
@@ -0,0 +1,256 @@
+          case MAP:
+            return visitMap(partnerType, group, visitor);
+          default:
+        }
+      }
+      return visitor.struct(partnerType, group, visitFields(partnerType, group, visitor));
+    }
+  }
+
+  private static <P, T> T visitList(P list, GroupType group, ParquetTypeWithPartnerVisitor<P, T> visitor) {
+    Preconditions.checkArgument(!group.isRepetition(Type.Repetition.REPEATED),
+        "Invalid list: top-level group is repeated: %s", group);
+    Preconditions.checkArgument(group.getFieldCount() == 1,
+        "Invalid list: does not contain single repeated field: %s", group);
+
+    GroupType repeatedElement = group.getFields().get(0).asGroupType();
+    Preconditions.checkArgument(repeatedElement.isRepetition(Type.Repetition.REPEATED),
+        "Invalid list: inner group is not repeated");
+    Preconditions.checkArgument(repeatedElement.getFieldCount() <= 1,
+        "Invalid list: repeated group is not a single field: %s", group);
+
+    visitor.beforeRepeatedElement(repeatedElement);
+    try {
+      T elementResult = null;
+      if (repeatedElement.getFieldCount() > 0) {
+        Type elementField = repeatedElement.getType(0);
+        try {
+          visitor.beforeElementField(elementField);
+          elementResult = visit(visitor.arrayElementType(list), elementField, visitor);
+        } finally {
+          visitor.afterElementField(elementField);
+        }
+      }
+      return visitor.list(list, group, elementResult);
+    } finally {
+      visitor.afterRepeatedElement(repeatedElement);
+    }
+  }
+
+  private static <P, T> T visitMap(P map, GroupType group, ParquetTypeWithPartnerVisitor<P, T> visitor) {
+    Preconditions.checkArgument(!group.isRepetition(Type.Repetition.REPEATED),
+        "Invalid map: top-level group is repeated: %s", group);
+    Preconditions.checkArgument(group.getFieldCount() == 1,
+        "Invalid map: does not contain single repeated field: %s", group);
+
+    GroupType repeatedKeyValue = group.getType(0).asGroupType();
+    Preconditions.checkArgument(repeatedKeyValue.isRepetition(Type.Repetition.REPEATED),
+        "Invalid map: inner group is not repeated");
+    Preconditions.checkArgument(repeatedKeyValue.getFieldCount() <= 2,
+        "Invalid map: repeated group does not have 2 fields");
+
+    visitor.beforeRepeatedKeyValue(repeatedKeyValue);
+    try {
+      T keyResult = null;
+      T valueResult = null;
+      switch (repeatedKeyValue.getFieldCount()) {
+        case 2:
+          // if there are 2 fields, both key and value are projected
+          Type keyType = repeatedKeyValue.getType(0);
+          visitor.beforeKeyField(keyType);
+          try {
+            keyResult = visit(visitor.mapKeyType(map), keyType, visitor);
+          } finally {
+            visitor.afterKeyField(keyType);
+          }
+          Type valueType = repeatedKeyValue.getType(1);
+          visitor.beforeValueField(valueType);
+          try {
```

Comment:
Maybe we can have a method `runWithStack(Runnable, Type)`?

Reply:
`ParquetTypeVisitor` also has the same visiting code logic. How about using a separated PR to refactor both?